plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiTableFunction.java

```diff
@@ -0,0 +1,151 @@
+                                .name("QUERY")
+                                .type(VarcharType.VARCHAR)
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("START")
+                                .type(TIMESTAMP_TZ_NANOS)
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("END")
+                                .type(TIMESTAMP_TZ_NANOS)
+                                .build()),
+                GENERIC_TABLE);
+
+        this.metadata = metadata;
+    }
+
+    @Override
+    public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments, ConnectorAccessControl accessControl)
+    {
+        var argument = (ScalarArgument) arguments.get("QUERY");
+        String query = ((Slice) argument.getValue()).toStringUtf8();
+
+        LongTimestampWithTimeZone startArgument = (LongTimestampWithTimeZone) ((ScalarArgument) arguments.get("START")).getValue();
+        LongTimestampWithTimeZone endArgument = (LongTimestampWithTimeZone) ((ScalarArgument) arguments.get("END")).getValue();
+
+        if (Strings.isNullOrEmpty(query)) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, query);
+        }
+        requireNonNull(startArgument);
+        requireNonNull(endArgument);
+        requireNonNull(metadata);
+
+        // determine the returned row type
+        List<ColumnHandle> columnHandles;
+        try {
+            columnHandles = metadata.getColumnHandles(query);
+        }
+        catch (UndeclaredThrowableException e) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Cannot get column definition", Throwables.getRootCause(e));
+        }
+        if (columnHandles.isEmpty()) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Cannot get column definition");
+        }
+        Descriptor returnedType = new Descriptor(columnHandles.stream()
+                .map(LokiColumnHandle.class::cast)
+                .map(column -> new Descriptor.Field(column.name(), Optional.of(column.type())))
+                .collect(toImmutableList()));
+
+        Instant end = toInstant(endArgument);
+        Instant start = toInstant(startArgument);
+
+        var tableHandle = new LokiTableHandle(
+                query,
+                start,
+                end,
+                columnHandles);
+
+        return TableFunctionAnalysis.builder()
+                .returnedType(returnedType)
+                .handle(new QueryHandle(tableHandle))
+                .build();
+    }
+
+    /**
+     * Convert LongTimestampWithTimeZone to Instant. See BigQuery code for an example
+     */
+    private Instant toInstant(LongTimestampWithTimeZone ts)
+    {
+        return Instant.ofEpochMilli(ts.getEpochMillis());
```

Comment:
Good point. I've copied the conversion from the [MySQL client](https://github.com/trinodb/trino/blob/c3d77ba39584f6ed2de973a19652c1db7ad8a990/plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java#L703-L705). I don't know the motivation for the different packed timestamps but it would be nice to have a `toInstant` method.